compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp

```diff
@@ -0,0 +1,449 @@
+ *
+ *  [after]  _shape :
+ *              - value :   (1)   (448)  (7)  (7)
+ *              - tags  :   (-)   (2)    (0)  (1)
+ */
+template <loco::DataType PermType>
+bool TaggedShapeAnalyzer::analyze_transpose(const luci::CircleTranspose *transpose_node)
+{
+  const luci::CircleConst *perm_node = loco::must_cast<luci::CircleConst *>(transpose_node->perm());
+  assert(perm_node->dtype() == PermType);
+
+  TaggedShapeAnalyzer::Shape new_shape;
+  const auto size = perm_node->size<PermType>();
+  for (uint32_t i = 0; i < size; i++)
+  {
+    auto perm_idx = perm_node->at<PermType>(i);
+    new_shape.push_back(_shape.at(perm_idx));
+  }
+  _shape = new_shape;
+  return true;
+}
+
+/**
+ * @brief update _shape based on 'Reshape' shape value
+ *
+ * @example Let's assume Reshape(shape=1, 448, 49) is given to [before] _shape.
+ *
+ *  [before]  _shape :
+ *              - value :   (1)   (448)  (7)   (7)
+ *              - tags  :   (-)   (2)    (0)   (1)
+ *
+ *  [after]  _shape :
+ *              - value :   (1)   (448)  (49)
+ *              - tags  :   (-)   (2)    (0, 1)
+ */
+template <loco::DataType ReshapeType>
+bool TaggedShapeAnalyzer::analyze_reshape(const luci::CircleReshape *reshape_node)
+{
+  const luci::CircleConst *shape_node = loco::must_cast<luci::CircleConst *>(reshape_node->shape());
+  assert(shape_node->dtype() == ReshapeType);
+
+  // At least one element must be in reshape's output-tensor.
+  if (shape_node->size<ReshapeType>() <= 0)
+    return false;
+
+  // Create new_shape based on reshape_node/shape
+  Shape new_shape;
+  for (uint32_t i = 0; i < shape_node->size<ReshapeType>(); i++)
+  {
+    TaggedShapeAnalyzer::Dim dim;
+    dim.value = shape_node->at<ReshapeType>(i);
+    new_shape.push_back(dim);
+  }
+
+  // indexing for _shape [old_shape_start_idx, old_shape_end_idx)
+  uint32_t old_shape_start_idx = 0;
+  uint32_t old_shape_end_idx = 1;
+  uint32_t old_shape_product = _shape[old_shape_start_idx].value;
+
+  auto expand_range = [&]() -> bool {
+    if (old_shape_end_idx >= _shape.size())
+      return false;
+
+    old_shape_product *= _shape[old_shape_end_idx].value;
+    old_shape_end_idx++;
+    return true;
+  };
+
+  auto move_to_next_range = [&]() -> bool {
+    if (old_shape_end_idx >= _shape.size())
+      return false;
+
+    old_shape_start_idx = old_shape_end_idx;
+    old_shape_end_idx++;
+    old_shape_product = _shape[old_shape_start_idx].value;
+    return true;
+  };
+
+  // Add tags from '_shape' to the 'new_shape'
+  uint32_t new_shape_idx = 0;
+  while (new_shape_idx < new_shape.size())
+  {
+    Dim &target_dim = new_shape[new_shape_idx];
+
+    // Ignore dim == 1
+    if (target_dim.value == 1)
+    {
+      new_shape_idx++;
+      continue;
+    }
+
+    while (old_shape_product < target_dim.value)
+    {
+      if (expand_range() == false)
+        break;
+    }
+
+    if (old_shape_product != target_dim.value)
+      return false;
+
+    // old_shape_product == target_dim.value
```

Comment:
Assertion would make it safer.
```suggestion
    assert (old_shape_product == target_dim.value);
```